Miscellaneous SuperIlc Improvements, pt# 3

trylek opened the pull request:
1) Add new JitRunner representing a no-op compilation step
followed by JIT-only execution of the source IL assembly.

2) CompilerRunner - add timeout constant for execution
with / without GC stress and support for environment overrides.

3) ParallelRunner - I have come to the conclusion that
in ETW mode we need to keep the process ID alive for
the entire duration of the [potentially parallel] ETW run,
otherwise OS may end up recycling some PIDs and we become
unable to back-translate them to logical execution processes.

To compensate for this somewhat ugly requirement, I have
added logic to split the execution process set to chunks
(currently each chunk corresponds to 1000 processes);
we only apply parallelism within a particular chunk,
the chunks are run in serial and each creates a new trace
event session and the ReadyToRunJittedMethods filter.

Thanks

Tomas

nattress commented on tests/src/tools/ReadyToRun.SuperIlc/CompileDirectoryCommand.cs:
```diff
@@ -57,6 +57,8 @@ class CompileDirectoryCommand
             }
 
             List<CompilerRunner> runners = new List<CompilerRunner>();
+            runners.Add(new JitRunner(null, inputDirectory.ToString(), outputDirectory.ToString(), referencePaths));
```
I don't see a class for this in the diff. Is there a JitRunner.cs that should be in the commit?

trylek replied:
Of course you're right, sorry about the overlooking, should be fixed in 2nd commit together with one other overlooking (I didn't actually add support for the environment overrides to ProcessRunner even though I added them to the ProcessInfo). I'm on the verge of being able to publish the per-assembly summary statistics table, I'll mention it in this PR thread once I get it.

trylek commented:
Hmm, it looks like the test assembly itself is for some reason missing in the JIT summary, I'm investigating why.

trylek commented:
Updated JIT statistics table: I have changed the ordering to CPAOT - Crossgen - JIT because in practice this acts as a form of partial ordering w.r.t. JITting the methods. As mentioned above, after being able to produce this kind of statistics I'm about to expand to multiple tests and to CPAOT-built framework.

<pre>
    CPAOT | Crossgen |      Jit | Assembly Name
-----------------------------------------------
        1 |        1 |      115 | ReadyToRunUnit
        6 |        6 |        6 | system.collections
        5 |        5 |        5 | system.collections.concurrent
        7 |        7 |        7 | system.collections.immutable
       37 |       37 |       37 | system.console
        6 |        6 |        6 | system.diagnostics.stacktrace
       10 |       10 |       10 | system.io.filesystem
        2 |        2 |        2 | system.linq
       50 |       50 |       50 | system.linq.expressions
       76 |       79 |     1492 | System.Private.CoreLib
       60 |      139 |      139 | system.reflection.metadata
</pre>

Thanks

Tomas

nattress commented:
This table looks great - thanks for implementing this. Did you run it against the ASP.NET functional test?

trylek commented:
Not yet by far, right now I'm still at the ReadyToRunUnit level. I'm slowly expanding but it takes time, hopefully to bring you some good news tomorrow ;-). I'm merging this in now as it passes the CI testing and it lets me reduce the set of my outstanding local changes.
